src/base/Gas.ml

```diff
@@ -104,70 +104,157 @@ module ScillaGas (SR : Rep) (ER : Rep) = struct
+                  ( o,
+                    List.map clauses ~f:(fun (p, e') ->
+                        (p, expr_static_cost e')) ),
+                erep ) )
+      | Builtin _
+      (* We don't add costs for Builtin because we can't know it statically
+       * without type info (Eval doesn't run the type checker). To know this
+       * statically, call builtin_cost below, providing argument types. *)
+      | GasExpr _ ->
+          ee
+    in
+    (e', erep)
 
   (* this is a dynamic cost. *)
+  let rec stmts_cost = function
+    | [] -> []
+    | (s, srep) :: rem_stmts ->
+        let s' =
+          match s with
+          | Load (x, _) ->
+              let g =
+                GasStmt
+                  (GasCharge.SumOf
+                     ( GasCharge.SizeOf (GI.get_id x),
+                       GasCharge.MapSortCost (GI.get_id x) ))
+              in
+              (* We charge *after* the load because we can't know the size before. *)
+              [ (s, srep); (g, srep) ]
+          | Store (_, v) | SendMsgs v | CreateEvnt v ->
+              let g = GasStmt (GasCharge.SizeOf (GI.get_id v)) in
+              [ (g, srep); (s, srep) ]
+          | Bind (x, e) ->
+            let g = GasStmt (GasCharge.StaticCost 1) in
+            let s' = Bind (x, expr_static_cost e) in
+            [ (g, srep); (s', srep) ]
+          | ReadFromBC _ | CallProc _ ->
+              let g = GasStmt (GasCharge.StaticCost 1) in
+              [ (g, srep); (s, srep) ]
+          | MapUpdate (_, klist, ropt) ->
+              let n = GasCharge.StaticCost (List.length klist) in
+              let g =
+                match ropt with
+                | Some r -> GasCharge.SumOf (GasCharge.SizeOf (GI.get_id r), n)
+                | None -> n
+              in
+              [ (GasStmt g, srep); (s, srep) ]
+          | MapGet (x, _, klist, _) ->
+              let n = GasCharge.StaticCost (List.length klist) in
+              let g =
+                  GasCharge.SumOf
+                    ( GasCharge.SumOf
+                        ( GasCharge.SizeOf (GI.get_id x),
+                          GasCharge.MapSortCost (GI.get_id x) ),
+                      n )
+              in
+              [ (s, srep); (GasStmt g, srep) ]
+          | MatchStmt (x, clauses) ->
+              let g = GasCharge.StaticCost (List.length clauses) in
+              let clauses' =
+                List.map clauses ~f:(fun (p, stmts) ->
+                    let stmts' = stmts_cost stmts in
+                    (p, stmts'))
+              in
+              let s' = MatchStmt (x, clauses') in
+              [ (GasStmt g, srep); (s', srep) ]
+          | AcceptPayment ->
+              let g = GasStmt (GasCharge.StaticCost 1) in
+              [ (g, srep); (s, srep) ]
+          | Iterate (l, _) ->
+              let g = GasStmt (GasCharge.LengthOf (GI.get_id l)) in
+              [ (g, srep); (s, srep) ]
+          | Throw _ (* TODO: Throw should charge same as event and send. *)
```

Comment:
This TODO needs doing.

Reply:
I wanted to keep the gas changes (in tests) minimal, retaining as much as possible current gas charges. I'll do this as a separate change right after we are done with this PR (immediately, so I won't create an Issue to track it).